### PR TITLE
Fix Log4j context attributes on AWS Lambda logger

### PR DESCRIPTION
**Pull request by pcholakov**
Our implementation of `supplyContextData()` was previously returning immutable maps but that was in conflict with the default implementation of `supplyStringMap()`:

```java
default StringMap supplyStringMap() {
    return new JdkMapAdapterStringMap(supplyContextData());
}
```

The result is runtime exceptions like these – even if we suppress them, an exception is still created which will be bad for performance.

```
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 WARN StatusConsoleListener org.apache.logging.log4j.spi.AbstractLogger caught java.lang.UnsupportedOperationException logging ReusableSimpleMessage: Closing state machine
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 java.lang.UnsupportedOperationException
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at java.base/java.util.ImmutableCollections.uoe(Unknown Source)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at java.base/java.util.ImmutableCollections$AbstractImmutableMap.clear(Unknown Source)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.impl.JdkMapAdapterStringMap.clear(JdkMapAdapterStringMap.java:122)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.impl.MutableLogEvent.clear(MutableLogEvent.java:143)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.impl.ReusableLogEventFactory.release(ReusableLogEventFactory.java:141)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.config.LoggerConfig.log(LoggerConfig.java:578)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.config.AwaitCompletionReliabilityStrategy.log(AwaitCompletionReliabilityStrategy.java:92)
2023-12-12T11:14:41.399000+00:00 2023/12/12/[18]34dbc67bfa374752b962d068a576efc6 at org.apache.logging.log4j.core.Logger.log(Logger.java:169)
```

This PR contains two related changes:

- remove the `GrpcContextDataProvider` custom log4j context provider altogether; this was the cause of the above problem and was causing other issues with custom attributes in the AWS Lambda environment
- update the base Lambda request handler to make use of this to store the AWS request ID in a conventional key name, which can be used by downstream log formatters - including the new AWS log4j Lambda logger implementation

--- 

### Kotlin HTTP handler

```
2023-12-13 16:59:50 DEBUG [greeter.Greeter/Greet] dev.restate.sdk.core.RestateGrpcServer - Start processing call to greeter.Greeter/Greet
2023-12-13 16:59:50 DEBUG [greeter.Greeter/Greet] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=WAITING_START, id=x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ} to REPLAYING
2023-12-13 16:59:50 DEBUG [greeter.Greeter/Greet][x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=REPLAYING, id=x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ} to PROCESSING
2023-12-13 16:59:50 INFO  [greeter.Greeter/Greet][x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ] dev.restate.sdk.examples.Greeter - Hello from log4j!
2023-12-13 16:59:50 DEBUG [greeter.Greeter/Greet][x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=PROCESSING, id=x3s6u2-HrNkAYxjsPNUcIKZI5ulnoxkUQ} to CLOSED
```

### Java HTTP handler

Note the missing operation name and invocation id from the "Hello from log4j" line:

```
2023-12-13 17:01:04 DEBUG [greeter.Greeter/Greet] dev.restate.sdk.http.vertx.RequestHttpServerHandler - Handling request to greeter.Greeter/Greet
2023-12-13 17:01:04 DEBUG [greeter.Greeter/Greet] dev.restate.sdk.core.RestateGrpcServer - Start processing call to greeter.Greeter/Greet
2023-12-13 17:01:04 DEBUG [greeter.Greeter/Greet] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=WAITING_START, id=x3s6u2-HrNkAYxjshUVdvy7gRsnmQwDog} to REPLAYING
2023-12-13 17:01:04 DEBUG [greeter.Greeter/Greet][x3s6u2-HrNkAYxjshUVdvy7gRsnmQwDog] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=REPLAYING, id=x3s6u2-HrNkAYxjshUVdvy7gRsnmQwDog} to PROCESSING
2023-12-13 17:01:04 INFO   dev.restate.sdk.examples.Greeter - Hello from log4j!
2023-12-13 17:01:04 DEBUG [greeter.Greeter/Greet][x3s6u2-HrNkAYxjshUVdvy7gRsnmQwDog] dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=PROCESSING, id=x3s6u2-HrNkAYxjshUVdvy7gRsnmQwDog} to CLOSED
```

### AWS Lambda

These are from an updated https://github.com/restatedev/examples/pull/50.

#### Plain text

```
2023-12-13T15:09:30.400000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 START RequestId: 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd Version: 72
2023-12-13T15:09:30.917000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:30 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [greeter.Greeter/Greet] DEBUG dev.restate.sdk.core.RestateGrpcServer - Start processing call to greeter.Greeter/Greet
 - customId=
2023-12-13T15:09:31.658000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:31 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [greeter.Greeter/Greet] DEBUG dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=WAITING_START, id=x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg} to REPLAYING
 - customId=
2023-12-13T15:09:33.136000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:33 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg][greeter.Greeter/Greet] DEBUG dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=REPLAYING, id=x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg} to PROCESSING
 - customId=
2023-12-13T15:09:33.936000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:33 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg][greeter.Greeter/Greet] INFO  dev.restate.sdk.examples.Greeter - Hello from log4j!
 - customId=23ff1ee1-116d-48f5-ba74-a0879e2b6d47
2023-12-13T15:09:34.537000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:34 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg][greeter.Greeter/Greet] DEBUG dev.restate.sdk.core.InvocationStateMachine - Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=PROCESSING, id=x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg} to CLOSED
 - customId=
2023-12-13T15:09:34.538000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 2023-12-13 15:09:34 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd [x3s6u2-HrNkAYxjuc2JfkeJlMmelKNnGg][greeter.Greeter/Greet] DEBUG dev.restate.sdk.core.InvocationStateMachine - Closing state machine
 - customId=
2023-12-13T15:09:34.576000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 END RequestId: 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd
2023-12-13T15:09:34.576000+00:00 2023/12/13/[72]0303138f718143848a4ff258c8e2d9c7 REPORT RequestId: 0a449b7c-54a8-4d6f-a6fc-89facdbdf9fd	Duration: 4176.41 ms	Billed Duration: 4177 ms	Memory Size: 128 MB	Max Memory Used: 111 MB
```

#### JSON format

```
2023-12-13T15:06:27.797000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:27.797Z","level":"DEBUG","message":"Start processing call to greeter.Greeter/Greet","logger":"dev.restate.sdk.core.RestateGrpcServer","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","restateInvocationId":"greeter.Greeter/Greet"}
2023-12-13T15:06:28.477000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:28.477Z","level":"DEBUG","message":"Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=WAITING_START, id=x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA} to REPLAYING","logger":"dev.restate.sdk.core.InvocationStateMachine","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","restateInvocationId":"greeter.Greeter/Greet"}
2023-12-13T15:06:30.675000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:30.598Z","level":"DEBUG","message":"Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=REPLAYING, id=x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA} to PROCESSING","logger":"dev.restate.sdk.core.InvocationStateMachine","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","restateInvocationId":"greeter.Greeter/Greet","restateInvocationStatus":"REPLAYING","restateServiceMethod":"x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA"}
2023-12-13T15:06:31.477000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:31.477Z","level":"INFO","message":"Hello from log4j!","logger":"dev.restate.sdk.examples.Greeter","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","customId":"3c2ea100-ebd8-4caf-88b8-d81f85a60b12","restateInvocationId":"greeter.Greeter/Greet","restateInvocationStatus":"PROCESSING","restateServiceMethod":"x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA"}
2023-12-13T15:06:32.258000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:32.257Z","level":"DEBUG","message":"Transitioning InvocationStateMachine{serviceName='greeter.Greeter', state=PROCESSING, id=x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA} to CLOSED","logger":"dev.restate.sdk.core.InvocationStateMachine","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","restateInvocationId":"greeter.Greeter/Greet","restateInvocationStatus":"PROCESSING","restateServiceMethod":"x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA"}
2023-12-13T15:06:32.258000+00:00 2023/12/13/[71]683bf357a9724671959aed52b6be74b0 {"timestamp":"2023-12-13T15:06:32.258Z","level":"DEBUG","message":"Closing state machine","logger":"dev.restate.sdk.core.InvocationStateMachine","AWSRequestId":"7ac82183-a598-4d01-b090-43e1d5bb1411","restateInvocationId":"greeter.Greeter/Greet","restateInvocationStatus":"CLOSED","restateServiceMethod":"x3s6u2-HrNkAYxjtwKCcfaZmrnp01aYVA"}
```

### CODE REVIEW

File: sdk-lambda/src/main/java/dev/restate/sdk/lambda/BaseRestateLambdaHandler.java [hunk 1]
```diff
@@ -41,6 +44,11 @@ protected BaseRestateLambdaHandler() {
   @Override
   public APIGatewayProxyResponseEvent handleRequest(
       APIGatewayProxyRequestEvent input, Context context) {
-    return restateLambdaEndpoint.handleRequest(input, context);
+    try {
+      ThreadContext.put(AWS_REQUEST_ID, context.getAwsRequestId());
```
Comment (slinkydeveloper) [inline sdk-lambda/src/main/java/dev/restate/sdk/lambda/BaseRestateLambdaHandler.java:48]: Maybe use https://logging.apache.org/log4j/2.x/manual/thread-context.html#closeablethreadcontext?

File: sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java [hunk 2]
```diff
@@ -24,17 +26,16 @@ public Map<String, String> supplyContextData() {
...
-      return Map.of();
-    } else if (invocationId == null) {
-      return Map.of(RestateGrpcServer.LoggingContextSetter.SERVICE_METHOD_KEY, serviceMethod);
-    } else {
-      return Map.of(
-          RestateGrpcServer.LoggingContextSetter.INVOCATION_ID_KEY,
-          invocationId.toString(),
-          RestateGrpcServer.LoggingContextSetter.SERVICE_METHOD_KEY,
-          serviceMethod);
+    var context = new HashMap<>(ThreadContext.getContext());
```
Comment (slinkydeveloper) [inline sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java:29]: I don't fully understand why we need to propagate the ThreadContext here. Will log4j2 just merge all the different context providers? https://logging.apache.org/log4j/2.x/manual/extending.html#custom-contextdataprovider

Can you also test this change works as expected with the http endpoint?

File: sdk-lambda/src/main/java/dev/restate/sdk/lambda/BaseRestateLambdaHandler.java (hunk 1 above)
Comment (pcholakov) [reply sdk-lambda/src/main/java/dev/restate/sdk/lambda/BaseRestateLambdaHandler.java:48]: TIL!

File: sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java (hunk 2 above)
Comment (pcholakov) [reply sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java:29]: I also read that page but it didn't answer the question of multiple providers co-existing.

In my testing, log4j did NOT merge the contexts – only the two keys above were available prior to my changes. Will re-test and confirm though.

Comment (slinkydeveloper) [reply sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java:29]: ok then let's go ahead with this but let's verify nothing breaks with http endpoints. After release let's re-evaluate this bit about logging context integration.

Comment (pcholakov) [reply sdk-core/src/main/java/dev/restate/sdk/core/GrpcContextDataProvider.java]: Removed altogether as discussed - thanks for your help! It only breaks for handler-emitted log entries from Java, but not from Kotlin, as expected.

**Review by slinkydeveloper — APPROVED** [review]
Ready to go. Squash merge it

**Comment by pcholakov** [comment]
Done! Opened a backlog issue to track the regression in Java handler logging: https://github.com/restatedev/sdk-java/issues/180.

**Pull request merged**